### DISCUSSION TITLE
Running mitmweb behind a reverse proxy

### DISCUSSION THREAD

**Discussion by Fabricio20** [mitmproxy/mitmproxy#7198] [General]
Hello! I have been trying to get mitmweb to run behind a reverse proxy (mostly so I can easily access it under `mitmproxy.local` in my lab environment) as I use it in wireguard mode however it looks like mitmweb has some weird issues when running behind a reverse proxy.

The following config is what currently works, however this has some nasty workarounds for bugs in mitmweb and i'll detail them below.
```nginx
server {
    listen 80;
    listen [::]:80;

    server_name mitmproxy.local;
    port_in_redirect off;

    location / {
        proxy_set_header    Host '127.0.0.1:8081'; // necessary otherwise it breaks (404: Not Found)!
        proxy_set_header    X-Real-IP $remote_addr;
        proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header    X-Forwarded-Proto $scheme;
        proxy_set_header    Origin 'http://127.0.0.1:8081/'; // necessary otherwise websocket breaks!

        proxy_set_header    X-Forwarded-Host $host;
        proxy_set_header    X-Forwarded-Port $server_port;
        proxy_pass          http://127.0.0.1:8081/;

        proxy_set_header    Upgrade $http_upgrade;
        proxy_set_header    Connection $http_connection;
        proxy_http_version  1.1; // necessary otherwise it breaks (infinite loading)!
    }
}
```

### First issue

While using a standard nginx reverse proxy setup (no workarounds) mitmweb is just going to reply with a 404 when trying to access `/` behind a reverse proxy. The cause seems to be that when running behind a reverse proxy, the `Host` header is set by the proxy to the domain being proxied (`mitmproxy.local`) however for some reason mitmweb tries to process that header and replies with a `404: Not Found` page instead of the default index page with the react scripts. Interesting sidenote but if you try to access anything other than the index (`/`) such as `/static/app.js` it works just fine, the only page that returns a 404 is the index/main page.

The workaround for this is to set the `Host` header to whatever the mitmproxy is bound to -> `127.0.0.1:8081`.

### Second issue

While using a standard nginx reverse proxy setup (no workarounds) mitmweb will not properly close connections after web requests, this seems to be caused by some http version mismatch or similar, as the page just loads forever, however under a debug connection you can see the requests were replied to. The workaround here is to tell nginx to proxy the connection under HTTP/1.1 via the `proxy_http_version` config option.

Not entirely sure if this is to blame on nginx or mitmweb, however this does not happen with other applications under this standard setup so it might be worth investigating, or possibly at least documenting!

### Third issue

Once you get the mitmweb page to render, websocket is not going to work even after correctly setting up websocket support on nginx. The web browser just gets a connection closed with code 1006. Upon further inspection via a tool such as Postman, mitmweb is actually returning a 403 to the websocket connection with an error "Cross origin requests are not supported for websocket".

So it seems mitmweb is doing some CORS verification on the websocket requests (but not http?) and since the `Origin` header set by the browser is going to be `http://mitmproxy.local/` instead of the `http://127.0.0.1:8081/` that it's expecting the connection just gets closed.

### Reproducing these issues

This discussion is mostly a way to document the process for users that are interested in setting this up, however if the mitmproxy team is willing to take a look at the issues, you can reproduce it with the following standard nginx reverse proxy config:
```nginx
server {
    listen 80;
    listen [::]:80;

    server_name mitmproxy.local;
    port_in_redirect off;

    location / {
        proxy_set_header    Host $host;
        proxy_set_header    X-Real-IP $remote_addr;
        proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header    X-Forwarded-Proto $scheme;
        proxy_set_header    X-Forwarded-Host $host;
        proxy_set_header    X-Forwarded-Port $server_port;
        proxy_pass              http://127.0.0.1:8081;
        proxy_set_header    Upgrade $http_upgrade;
        proxy_set_header    Connection $http_connection;
    }
}
```

While debugging these issues, I came across previous discussion under https://github.com/mitmproxy/mitmproxy/discussions/6960 however it's not clear how that setup worked due to the above mentioned issues, I also came across https://github.com/mitmproxy/mitmproxy/issues/4963 as a general support request and https://github.com/mitmproxy/mitmproxy/issues/3393 which might be a more permanent fix for the websocket issue (being worked under https://github.com/mitmproxy/mitmproxy/pull/6443.

**Comment by 45gfg9** [comment]
Hands up on this: I am able to make it work using Caddy, but a certain feature is broken.

My setup: a NAS server that I can connect using Tailscale. mitmproxy running in container. Caddy as web server, configuration as below:
```
handle_path /mitm/* {
    reverse_proxy http://127.0.0.1:20138 {
        header_up Host 127.0.0.1:8081
        header_up Origin http://127.0.0.1:8081
    }
}
```
This works for web UI that I can access at `https://my-device/mitm/`, but is broken when I try to save traffic. It redirects to `https://my-device/flows/dump` which is not handled by itself. I have to manually access `/mitm/flows/dump` to dump the traffic.
